crates/sui-framework/packages/move-stdlib/sources/internal.move

```diff
@@ -0,0 +1,42 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Defines the `Permit` type, which can be used to constrain the logic of a
+/// generic function to be authorized only by the module that defines the type
+/// parameter.
+///
+/// ```move
+/// module example::use_permit;
+///
+/// use std::internal;
+///
+/// public struct MyType { /* ... */ }
+///
+/// public fun test_permit() {
+///    let permit = internal::permit<MyType>();
+///    /* external_module::call_with_permit(permit); */
+/// }
+/// ```
+///
+/// To write a function that is guarded by a `Permit`, require it as an argument.
+///
+/// ```move
+/// // Silly mockup of a type registry where a type can be registered only by
+/// // the module that defines the type.
+/// module example::type_registry;
+///
+/// use std::internal;
+///
+/// public fun register_type<T>(_: internal::Permit<T> /* ... */) {
```

Comment:
It can be another PR, but we should add automatic aliases for `std::internal as internal`